spec/java/util/ArrayList.main.lsl

```diff
@@ -0,0 +1,881 @@
+        {
+            val expectedModCount: int = this.modCount;
+
+            // Java has no unsigned primitive data types
+            action ASSUME(this.length > 0);
+
+            // plain bubble sorting algorithm with no optimizations
+            var i: int = 0;
+            var j: int = 0;
+            action LOOP_FOR(
+                i, 0, this.length, +1,
+                sort_loop_outer(i, j, c)
+            );
+
+            _checkForComodification(expectedModCount);
+        }
+
+        this.modCount += 1;
+    }
+
+    @Phantom proc sort_loop_outer (i: int, j: int, c: Comparator): void
+    {
+        action LOOP_FOR(
+            j, 0, this.length, +1,
+            sort_loop_inner(i, j, c)
+        );
+    }
+
+    @Phantom proc sort_loop_inner (i: int, j: int, c: Comparator): void
+    {
+        val a: Object = action LIST_GET(this.storage, i);
+        val b: Object = action LIST_GET(this.storage, j);
+
+        if (action CALL(c, [a, b]) > 0)
+        {
+            action LIST_SET(this.storage, i, b);
+            action LIST_SET(this.storage, j, a);
+        }
+    }
+
+
+    fun *.spliterator (@target self: ArrayList): Spliterator
+    {
+        result = new ArrayList_SpliteratorAutomaton(state = Initialized,
+            parent = self,
+        );
+    }
+
+
+    // within java.util.Collection
+    fun *.stream (@target self: ArrayList): Stream
+    {
+        // #problem: no streams
+        result = action SYMBOLIC("java.util.stream.Stream");
+        action ASSUME(result != null);
+    }
+
+
+    fun *.subList (@target self: ArrayList, fromIndex: int, toIndex: int): List
+    {
+        _subListRangeCheck(fromIndex, toIndex, this.length);
+
+        result = new ArrayList_SubListAutomaton(state = Initialized,
+            root = self,
+            parentList = null,
+            offset = fromIndex,
+            length = toIndex - fromIndex,
+            modCount = this.modCount,
+        );
+    }
+
+
+    fun *.toArray (@target self: ArrayList): array<Object>
+    {
+        val len: int = this.length;
+        result = action ARRAY_NEW("java.lang.Object", len);
+
+        var i: int = 0;
+        action LOOP_FOR(
+            i, 0, len, +1,
+            toArray_loop(i, result)
+        );
+    }
+
+    // #problem/todo: use exact parameter names
+    @Phantom proc toArray_loop (i: int, result: array<Object>): void
+    {
+        result[i] = action LIST_GET(this.storage, i);
+    }
+
+
+    // within java.util.Collection
+    fun *.toArray (@target self: ArrayList, generator: IntFunction): array<Object>
+    {
+        // acting just like the JDK: trigger NPE and class cast exceptions on invalid generator return value
+        val a: array<Object> = action CALL_METHOD(generator, "apply", [0]) as array<Object>;
+        val aLen: int = action ARRAY_SIZE(a);
+
+        val len: int = this.length;
+        // #problem: a.getClass() should be called to construct a type-valid array (USVM issue)
```

Comment:
This will crash with `ClassCastException` but should not:
```java
var list = new ArrayList<Integer>();
list.add(5);
Integer[] arr = list.toArray(new Integer[0]);
```